message_passing/waiting_for_handlers_and_compensation/workflows.py

```diff
@@ -0,0 +1,180 @@
+        # application logic; if this future completes before the message handler
+        # task then the handler should abort and perform compensation.
+        self.workflow_exit: asyncio.Future[None] = asyncio.Future()
+
+        # The following two attributes are implementation detail of this sample
+        # and can be ignored
+        self._update_started = False
+        self._update_compensation_done = False
+        self._workflow_compensation_done = False
+
+    @workflow.run
+    async def run(self, input: WorkflowInput) -> str:
+        try:
+            # 👉 Use this `try...except` style, instead of waiting for message
+            # handlers to finish in a `finally` block. The reason is that some
+            # exception types cause a workflow task failure as opposed to
+            # workflow exit, in which case we do *not* want to wait for message
+            # handlers to finish.
+
+            # 👉 self._run contains your actual application logic. This is
+            # implemented in a separate method in order to separate
+            # "platform-level" concerns (waiting for handlers to finish and
+            # ensuring that compensation is performed when appropriate) from
+            # application logic. In this sample, its actual implementation is
+            # below but contains nothing relevant.
+            result = await self._run(input)
+            self.workflow_exit.set_result(None)
+            await workflow.wait_condition(workflow.all_handlers_finished)
+            return result
+        # 👉 Catch BaseException since asyncio.CancelledError does not inherit
+        # from Exception.
+        except BaseException as e:
+            if is_workflow_exit_exception(e):
+                self.workflow_exit.set_exception(e)
+                await workflow.wait_condition(workflow.all_handlers_finished)
+                await self.workflow_compensation()
+                self._workflow_compensation_done = True
+            raise
+
+    async def workflow_compensation(self):
+        await workflow.execute_activity(
+            activity_executed_to_perform_workflow_compensation,
+            start_to_close_timeout=timedelta(seconds=10),
+        )
+        self._update_compensation_done = True
+
+    @workflow.update
+    async def my_update(self) -> str:
+        """
+        An update handler that handles exceptions raised in its own execution
+        and in that of the main workflow method.
+
+        It ensures that:
+        - Compensation/cleanup is always performed when appropriate
+        - The update caller gets the update result, or WorkflowUpdateFailedError
+        """
+        # 👉 As with the main workflow method, the update application logic is
+        # implemented in a separate method in order to separate "platform-level"
+        # error-handling and compensation concerns from application logic. Note
+        # that coroutines must be wrapped in tasks in order to use
+        # workflow.wait.
```

Comment:
nit: This struck me as abstract.  Lines 97-102 are what helped me understand what's going on, so I would probably drop this.

Reply:
Agreed, done.